### PR TITLE
Add ProcessRunner and PIDFile

### PR DESCRIPTION
**Pull request by pavledragisic**

### CODE REVIEW

File: Foundation/include/Poco/ProcessRunner.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+#include <vector>
+
+
+namespace Poco {
+
+
+class Foundation_API ProcessRunner: public Poco::Runnable
+	/// ProcessRunner is a wrapper class for `Poco::ProcessHandle.
+	/// It starts and terminates a process with enabled or disabled (default)
+	/// `stdio` pipes, and waits for process PID to be created before
```
Comment (aleks-f) [inline Foundation/include/Poco/ProcessRunner.h:39]: change to: and **optionally** waits for process PID to be created before

File: Foundation/src/ProcessRunner.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		// https://github.com/pocoproject/poco/blob/ed43543113c5307e216615d8c004f5ffd8a8b37f/Foundation/include/Poco/Process.h#L252
+		// the solution is either to implement Windows SIGTERM handling in Poco:
+		// https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/signal?view=msvc-170
+		// or resort to brute force TerminateProcess Win API for non-Poco processes:
+		// https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess
+		ProcessHandle* pPH = nullptr;
+		if ((pPH = _pPH.exchange(nullptr)))
+			Process::requestTermination(pPH->id());
+
+		std::string msg = "Waiting for process shutdown";
```
Comment (aleks-f) [inline Foundation/src/ProcessRunner.cpp:155]: messages are messed up here, this one belongs to line 171

File: Foundation/src/ProcessRunner.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		if (!_pidFile.empty())
+		{
+			Poco::format(msg, "%s (pidFile: '%s')", msg, _pidFile);
```
Comment (aleks-f) [inline Foundation/src/ProcessRunner.cpp:158]: this one should first set msg to "Waiting for PID file", then format it with the name

File: Foundation/src/ProcessRunner.cpp [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+			if (!_pidFile.empty())
+			{
+				File pidFile(_pidFile);
+				_pidFile.clear();
+				Stopwatch sw; sw.start();
+				while (pidFile.exists())
+					checkTimeout(sw, "waiting for PID file");
```
Comment (aleks-f) [inline Foundation/src/ProcessRunner.cpp:166]: msg should be used here

File: Foundation/src/ProcessRunner.cpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		bool startProcess,
+		const Args& pidArgFmt): _cmd(cmd),
+			_args(args),
+			_pid(INVALID_PID),
+			_pidFile(pidFile),
+			_options(options),
+			_timeout(timeout),
+			_pPH(nullptr),
+			_started(false),
+			_logger(Poco::Logger::get("ProcessRunner"))
```
Comment (aleks-f) [inline Foundation/src/ProcessRunner.cpp:53]: remove Logger, typically framework classes don't have logger, we communicate important things through exceptions

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/src/ProcessRunner.cpp (hunk 5 above)
Comment (pavledragisic) [reply Foundation/src/ProcessRunner.cpp]: Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)

File: Foundation/src/ProcessRunner.cpp (hunk 4 above)
Comment (pavledragisic) [reply Foundation/src/ProcessRunner.cpp]: Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)

File: Foundation/src/ProcessRunner.cpp (hunk 3 above)
Comment (pavledragisic) [reply Foundation/src/ProcessRunner.cpp:145]: Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)

File: Foundation/src/ProcessRunner.cpp (hunk 2 above)
Comment (pavledragisic) [reply Foundation/src/ProcessRunner.cpp]: Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)

File: Foundation/include/Poco/ProcessRunner.h (hunk 1 above)
Comment (pavledragisic) [reply Foundation/include/Poco/ProcessRunner.h]: Fixed in [ca79ba8](https://github.com/pocoproject/poco/pull/4225/commits/ca79ba8cb058dba942dd9b86e912037325339d6e)

File: Foundation/src/ProcessRunner.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	delete pPH;
+}
+
+
+void ProcessRunner::stop()
+{
+	if (_started)
+	{
+		// TODO:
```
Comment (aleks-f) [inline Foundation/src/ProcessRunner.cpp:131]: Ignore this for now, I was mistaken. I may make sense to implement `signal` on windows, but I'm not sure. Remove the coment for release.

File: Foundation/testsuite/src/ProcessRunnerTest.cpp [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+#ifdef _DEBUG
+	name += 'd';
+#endif
+
+	cmd = name;
+
+	// non-auto start, no PID
+	{
+		std::vector<std::string> args;
+		ProcessRunner pr(cmd, args, "", ProcessRunner::NO_OUT, 10, false);
```
Comment (aleks-f) [inline Foundation/testsuite/src/ProcessRunnerTest.cpp:136]: This works randomly, mostly not. The reason is that you send no command line arguments, the `TestApp` does not wait, completes before `ProcessRunner` can see a valid handle (remember, the `ProcessRunner::run()` is another thread), leaving it trying until it times out and the show is over.

File: Foundation/testsuite/src/ProcessRunnerTest.cpp [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+#include "Poco/Format.h"
+#include "Poco/Path.h"
+#include "Poco/File.h"
+#include "Poco/FileStream.h"
+#include "Poco/Environment.h"
+#include <fstream>
+#include <iostream>
+
+
+//using namespace IoT::Devs::Util;
```
Comment (aleks-f) [inline Foundation/testsuite/src/ProcessRunnerTest.cpp:27]: remove

File: Foundation/testsuite/src/ProcessRunnerTest.cpp [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+using namespace Poco;
+using namespace Poco::Util;
+//using IoT::Devs::Util::enableLogging;
```
Comment (aleks-f) [inline Foundation/testsuite/src/ProcessRunnerTest.cpp:30]: remove

File: Foundation/testsuite/src/TestApp.cpp [hunk 10]
```diff
@@ -57,6 +60,32 @@ int main(int argc, char** argv)
...
+			std::unique_ptr<PIDFile> _pPIDFile;
+			size_t equals_pos = arg.find('=');
+
+			if (equals_pos != std::string::npos)
+			{
+				std::string pidPath = arg.substr(equals_pos + 1);
+				_pPIDFile.reset(new PIDFile(pidPath, true, false));
+				_pPIDFile->create(false);
+
+				sigset_t sset;
```
Comment (aleks-f) [inline Foundation/testsuite/src/TestApp.cpp:74]: this is POSIX-only, need at least windows implementation, too. #ifdef out the complete `ProcessRunner` code out for android (@zaleksandar should try at least android build/testrun before relase to make sure we did not break something there)

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/testsuite/src/ProcessRunnerTest.cpp [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+	// non-auto start, no PID
+	{
+		std::vector<std::string> args;
+		ProcessRunner pr(cmd, args, "", ProcessRunner::NO_OUT, 10, false);
+		assertTrue (pr.cmdLine() == cmdLine(cmd, args));
+		assertFalse (pr.running());
+		pr.start();
+		Stopwatch sw; sw.start();
+		while (!pr.running())
+			checkTimeout(sw, "Waiting for process to start", 1000);
```
Comment (aleks-f) [inline Foundation/testsuite/src/ProcessRunnerTest.cpp:142]: make this `checkTimeout(sw, "Waiting for process to start", 1000, __LINE__);`,
and this:
https://github.com/pocoproject/poco/blob/95a107b5c0f7a1b2b987433b0a51db1a27eac7ca/Foundation/testsuite/src/ProcessRunnerTest.cpp#L303

into this:

```c++
void ProcessRunnerTest::checkTimeout(const Stopwatch& sw, const std::string& msg, int timeoutMS, int line)
{
	if (sw.elapsedSeconds()*1000 > timeoutMS)
	{
		throw Poco::TimeoutException(
			Poco::format("ProcessRunner::checkTimeout(): %s, line: %d", msg, line));
	}
	Thread::sleep(10);
}
```

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/include/Poco/PIDFile.h [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+{
+public:
+	using Ptr = std::unique_ptr<PIDFile>;
+
+	static const int INVALID_PID = -1;
+
+	PIDFile();
+		/// Creates the PIDFile.
+
+	PIDFile(const std::string& fileName, bool append = true, bool write = true);
```
[the remark this answers is not in the document]
Comment (pavledragisic) [reply Foundation/include/Poco/PIDFile.h]: Fixed in https://github.com/pocoproject/poco/pull/4225/commits/b02c2cc267557c0945d24709cddaced0e720e808

File: Foundation/testsuite/src/ProcessRunnerTest.cpp (hunk 7 above)
Comment (pavledragisic) [reply Foundation/testsuite/src/ProcessRunnerTest.cpp:104]: Should be fixed in https://github.com/pocoproject/poco/pull/4225/commits/f417b4c8981ad9886383befeb82cae4c18b2430b

File: Foundation/testsuite/src/ProcessRunnerTest.cpp (hunk 8 above)
Comment (pavledragisic) [reply Foundation/testsuite/src/ProcessRunnerTest.cpp]: Fixed in [b02c2cc](https://github.com/pocoproject/poco/pull/4225/commits/b02c2cc267557c0945d24709cddaced0e720e808)

File: Foundation/testsuite/src/ProcessRunnerTest.cpp (hunk 9 above)
Comment (pavledragisic) [reply Foundation/testsuite/src/ProcessRunnerTest.cpp]: Fixed in [b02c2cc](https://github.com/pocoproject/poco/pull/4225/commits/b02c2cc267557c0945d24709cddaced0e720e808)

File: Foundation/testsuite/src/ProcessRunnerTest.cpp (hunk 11 above)
Comment (pavledragisic) [reply Foundation/testsuite/src/ProcessRunnerTest.cpp]: Fixed in https://github.com/pocoproject/poco/pull/4225/commits/a08f5b5bb749a8983dcbdde145d9d4e398caebf3

File: Foundation/include/Poco/PIDFile.h [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+	~PIDFile();
+		/// Destroys the PIDFile.
+		/// If fileName is not empty, deletes the PID file.
+
+	const std::string& getName() const;
+		/// Returns the file name.
+
+	void setName(const std::string& fileName);
+		/// Sets the file name;
```
Comment (aleks-f) [inline Foundation/include/Poco/PIDFile.h:54]: comments should not end with `;` - besides being grammatically incorrect, it causes `CodeQL` to create alerts

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/include/Poco/ProcessRunner.h [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+	static const int NO_OUT = Poco::PROCESS_CLOSE_STDOUT|Poco::PROCESS_CLOSE_STDERR;
+		/// Constant to prevent std out and err from being received from the process.
+
+	ProcessRunner(const std::string& cmd,
+		const Args& args,
+		const std::string& pidFile = "",
+		int options = NO_OUT,
+		int timeout = 10, /*seconds*/
+		bool startProcess = true,
+		const Args& pidArgFmt = {"-p", "--pidfile="});
```
Comment (aleks-f) [inline Foundation/include/Poco/ProcessRunner.h:60]: the comment that was here still applies, windows has `/` convention for parameters; when I said "disregard windows" I meant for terminating the process, see here:
https://github.com/pocoproject/poco/blob/83d50284e07df356dcb387165d23bcc2b544682d/Util/src/HelpFormatter.cpp#L276-L293

**Review by aleks-f — CHANGES_REQUESTED** [review]

File: Foundation/include/Poco/PIDFile.h (hunk 13 above)
Comment (pavledragisic) [reply Foundation/include/Poco/PIDFile.h]: Fixed in [1b1feaf](https://github.com/pocoproject/poco/pull/4225/commits/1b1feaf5b4abf5f765e5a9cc656e70deb9b0813c)

File: Foundation/include/Poco/ProcessRunner.h (hunk 14 above)
Comment (pavledragisic) [reply Foundation/include/Poco/ProcessRunner.h]: Fixed in https://github.com/pocoproject/poco/pull/4225/commits/ce8ba27335a3d558c667a6c89c9f242019aaf22b

File: Foundation/src/ProcessRunner.cpp [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+		if (!_pidFile.empty())
+		{
+			if (!_pidFile.empty())
+			{
+				File pidFile(_pidFile);
+				_pidFile.clear();
+				std::string msg;
+				Poco::format(msg, "Waiting for PID file (pidFile: '%s')", _pidFile);
+				Stopwatch sw; sw.start();
```
[the remark this answers is not in the document]
Comment (pavledragisic) [reply Foundation/src/ProcessRunner.cpp]: Fixed in https://github.com/pocoproject/poco/pull/4225/commits/def4dbecc7e34aca1605388b460c7f9ea806dc70

File: Foundation/testsuite/src/TestApp.cpp (hunk 10 above)
Comment (pavledragisic) [reply Foundation/testsuite/src/TestApp.cpp]: Implemented in https://github.com/pocoproject/poco/pull/4225/commits/731d77093525dc9db472cd73b9b66d3c1ade11ec

**Review by aleks-f — APPROVED** [review]

**Comment by aleks-f** [comment]
@pavledragisic is this ready to merge?

**Comment by pavledragisic** [comment]
@aleks-f it's ready to merge.

**Review by aleks-f — APPROVED** [review]

**Pull request merged**